File: ui/article-aem-renderer/src/test/java/org/cru/godtools/article/aem/db/TranslationRepositoryIT.java
package org.cru.godtools.article.aem.db;

import android.net.Uri;

import org.cru.godtools.article.aem.model.AemImport;
import org.cru.godtools.article.aem.model.TranslationRef;
import org.cru.godtools.article.aem.model.TranslationRefKt;
import org.cru.godtools.model.Translation;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.annotation.Config;

import java.util.Arrays;
import java.util.Locale;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.robolectric.annotation.Config.ALL_SDKS;

@RunWith(AndroidJUnit4.class)
@Config(sdk = ALL_SDKS)
public class TranslationRepositoryIT extends BaseArticleRoomDatabaseIT {
    private static final String TOOL = "kgp";
    private static final Locale LANG = Locale.ENGLISH;
    private static final Uri URI1 = Uri.parse("https://example.com/content/experience-fragments/questions_about_god");
    private static final Uri URI2 = Uri.parse("https://example.com/content/experience-fragments/other");

    private Translation mTranslation;
    private TranslationRef.Key mTranslationKey;

    private TranslationRepository mRepository;

    @Before
    public final void setup() {
        mRepository = mDb.translationRepository();

        mTranslation = new Translation();
        mTranslation.setToolCode(TOOL);
        mTranslation.setLanguageCode(LANG);
        mTranslation.setVersion(1);
        mTranslationKey = TranslationRefKt.toTranslationRefKey(mTranslation);
    }

    @Test
    public void verifyAddAemImportsTranslationAlreadyPresent() {
        // setup test
        mDb.translationDao().insertOrIgnore(new TranslationRef(mTranslationKey));

        // perform test
        assertFalse(mRepository.isProcessed(mTranslation));
        mRepository.addAemImports(mTranslation, Arrays.asList(URI1, URI2));
        assertTrue(mRepository.isProcessed(mTranslation));

        // TODO: test AemImports once we define dao methods for reading AemImports

        AemImport aemImport = mDb.aemImportDao().find(URI1);
        assertEquals(URI1, aemImport.getUri());
    }
}

File: ui/article-aem-renderer/src/test/kotlin/org/cru/godtools/article/aem/db/TranslationRepositoryIT.kt
package org.cru.godtools.article.aem.db

import android.net.Uri
import androidx.test.ext.junit.runners.AndroidJUnit4
import java.util.Locale
import kotlinx.coroutines.runBlocking
import org.ccci.gto.android.common.androidx.room.RoomDatabaseRule
import org.cru.godtools.article.aem.model.TranslationRef
import org.cru.godtools.article.aem.model.toTranslationRefKey
import org.cru.godtools.model.randomTranslation
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

private val URI1 = Uri.parse("https://example.com/content/experience-fragments/questions_about_god")
private val URI2 = Uri.parse("https://example.com/content/experience-fragments/other")

@RunWith(AndroidJUnit4::class)
class TranslationRepositoryIT {
    @get:Rule
    internal val dbRule = RoomDatabaseRule(ArticleRoomDatabase::class.java)
    private val db get() = dbRule.db
    private val repository get() = dbRule.db.translationRepository()

    @Test
    fun verifyAddAemImportsTranslationAlreadyPresent() = runBlocking {
        // setup test
        val translation = randomTranslation(
            toolCode = "kgp",
            languageCode = Locale.ENGLISH,
            version = 1,
        )
        db.translationDao().insertOrIgnore(TranslationRef(translation.toTranslationRefKey()!!))

        // perform test
        assertFalse(repository.isProcessed(translation))
        repository.addAemImports(translation, listOf(URI1, URI2))
        assertTrue(repository.isProcessed(translation))

        // TODO: test AemImports once we define dao methods for reading AemImports
        val aemImport = runBlocking { db.aemImportDao().find(URI1) }
        assertEquals(URI1, aemImport!!.uri)
    }
}
